src/JuMP.jl

```diff
@@ -123,50 +174,108 @@ mutable struct Model <: AbstractModel
+                 Dict{Symbol, Any}())
+end
+
+"""
+    Model(; caching_mode::MOIU.CachingOptimizerMode=MOIU.Automatic,
+            bridge_constraints::Bool=true)
+
+Return a new JuMP model without any optimizer storing the model in a cache.
+The mode of the `CachingOptimizer` storing this cache is `caching_mode`.
+The optimizer can be set later with [`set_optimizer`](@ref). If
+`bridge_constraints` is true, constraints that are not supported by the
+optimizer are automatically bridged to equivalent supported constraints when
+an appropriate is defined in the `MathOptInterface.Bridges` module or is
+defined in another module and is explicitely added.
+"""
+function Model(; caching_mode::MOIU.CachingOptimizerMode=MOIU.Automatic,
+                 bridge_constraints::Bool=true)
+    universal_fallback = MOIU.UniversalFallback(JuMPMOIModel{Float64}())
+    caching_opt = MOIU.CachingOptimizer(universal_fallback,
+                                        caching_mode)
+    if bridge_constraints
+        backend = MOI.Bridges.fullbridgeoptimizer(caching_opt,
+                                                  Float64)
+    else
+        backend = caching_opt
     end
+    return Model(nothing, backend)
+end
+
+"""
+    Model(factory::OptimizerFactory;
+          caching_mode::MOIU.CachingOptimizerMode=MOIU.Automatic,
+          bridge_constraints::Bool=true)
+
+Return a new JuMP model using the factory `factory` to create the optimizer.
+This is equivalent to calling `Model` with the same keyword arguments and then
+calling [`set_optimizer`](@ref) on the created model with the `factory`. The
+factory can be created by the [`with_optimizer`](@ref) function.
+
+## Examples
+
+The following creates a model using the optimizer
+`IpoptOptimizer(print_level=0)`:
+```julia
+model = JuMP.Model(with_optimizer(IpoptOptimizer, print_level=0))
+```
+"""
+function Model(factory::OptimizerFactory; kwargs...)
+    model = Model(; kwargs...)
+    set_optimizer(model, factory)
+    return model
+end
+
+"""
+    direct_model(backend::MOI.ModelLike)
+
+Return a new JuMP model using `backend` to store the model and solve it. As
+opposed to the [`Model`](@ref) constructor, no cache of the model is stored
+outside of `backend` and no bridges are automatically applied to `backend`.
+The absence of cache reduces the memory footprint but it is importnat to bear
+in mind the following implications of creating models using this *direct* mode:
+
+* When `backend` does not support an operation such as adding
+  variables/constraints after solver or modifying constraints, an error is
+  thrown. With models created using the [`Model`](@ref) constructor, such
+  situations can be dealt with modifying the cache only and copying the model
+  cache once `JuMP.optimize` is called.
+* When `backend` does not support a constraint type, the constraint is not
+  automatically bridged to constraints supported by `backend`.
+* The optimizer used cannot be changed. With models created using the
```

Comment:
The explanation past "The optimizer used cannot be changed" is too verbose. I would just say "The optimizer cannot be changed after the Model is constructed."